test/jdk/java/net/httpclient/SpecialHeadersTest.java

```diff
@@ -140,6 +157,142 @@ public Object[][] variants() {
+        }
+
+
+    }
+
+    protected boolean stopAfterFirstFailure() {
+        return Boolean.getBoolean("jdk.internal.httpclient.debug");
+    }
+
+    final AtomicReference<SkipException> skiptests = new AtomicReference<>();
+    void checkSkip() {
+        var skip = skiptests.get();
+        if (skip != null) throw skip;
+    }
+    static String name(ITestResult result) {
+        var params = result.getParameters();
+        return result.getName()
+                + (params == null ? "()" : Arrays.toString(result.getParameters()));
+    }
+
+    @BeforeMethod
+    void beforeMethod(ITestContext context) {
+        if (stopAfterFirstFailure() && context.getFailedTests().size() > 0) {
+            if (skiptests.get() == null) {
+                SkipException skip = new SkipException("some tests failed");
+                skip.setStackTrace(new StackTraceElement[0]);
+                skiptests.compareAndSet(null, skip);
+            }
+        }
+    }
+
+    @AfterClass
+    static final void printFailedTests(ITestContext context) {
+        out.println("\n=========================");
+        var failed = context.getFailedTests().getAllResults().stream()
+                .collect(Collectors.toMap(r -> name(r), ITestResult::getThrowable));
+        FAILURES.putAll(failed);
+        try {
+            out.printf("%n%sCreated %d servers and %d clients%n",
+                    now(), serverCount.get(), clientCount.get());
+            if (FAILURES.isEmpty()) return;
+            out.println("Failed tests: ");
+            FAILURES.entrySet().forEach((e) -> {
+                out.printf("\t%s: %s%n", e.getKey(), e.getValue());
+                e.getValue().printStackTrace(out);
+            });
+            if (tasksFailed) {
+                out.println("WARNING: Some tasks failed");
+            }
+        } finally {
+            out.println("\n=========================\n");
+        }
+    }
+
+    private HttpClient makeNewClient() {
+        clientCount.incrementAndGet();
+        return HttpClient.newBuilder()
+                .proxy(NO_PROXY)
+                .executor(executor)
+                .sslContext(sslContext)
+                .build();
+    }
+
+    static volatile String lastMethod;
+    HttpClient newHttpClient(String method, boolean share) {
+        if (!share) return TRACKER.track(makeNewClient());
+        HttpClient shared = sharedClient;
+        String last = lastMethod;
+        if (shared != null && Objects.equals(last, method)) return shared;
+        synchronized (this) {
+            shared = sharedClient;
+            last = lastMethod;
+            if (!Objects.equals(last, method)) {
+                // reset sharedClient to avoid side effects
+                // between methods. This is needed to keep the test
+                // expectation that the first HTTP/2 clear request
+                // will be an upgrade
+                if (shared != null) {
+                    TRACKER.track(shared);
```

Comment:
Is it intentional that we track a shared client when it is being reset, instead of tracking it when we create it (a few lines later)? This is unlike a non-shared instance which we track when we create one.

Reply:
Yes - because the tracker will complain if any of the tracked client is still alive when check() is called. So the idea is to start tracking the shared client when we're no longer using it.